diff --git a/src/filelock/_unix.py b/src/filelock/_unix.py
index 4ae1fbe..dbbe408 100644
--- a/src/filelock/_unix.py
+++ b/src/filelock/_unix.py
@@ -26,7 +26,8 @@ if sys.platform == "win32":  # pragma: win32 cover
 else:  # pragma: win32 no cover
     try:
         import fcntl
-    except ImportError:
+        _ = (fcntl.lock, fcntl.LOCK_EX, fcntl.LOCK_NB, fcntl.LOCK_UN)
+    except (ImportError, AttributeError):
         pass
     else:
         has_fcntl = True
