diff --git a/IPython/__init__.py b/IPython/__init__.py
index b7235481f..679ee70d8 100644
--- a/IPython/__init__.py
+++ b/IPython/__init__.py
@@ -19,6 +19,8 @@
 # Imports
 #-----------------------------------------------------------------------------
 
+import jupyterlite_preload
+
 import sys
 
 #-----------------------------------------------------------------------------
diff --git a/IPython/core/interactiveshell.py b/IPython/core/interactiveshell.py
index 12c120625..76a5671fd 100644
--- a/IPython/core/interactiveshell.py
+++ b/IPython/core/interactiveshell.py
@@ -680,6 +680,9 @@ def __init__(self, ipython_dir=None, profile_dir=None,
         # `ipykernel.kernelapp`.
         self.trio_runner = None
 
+        import jupyterlite_preload
+        jupyterlite_preload._patch()
+
     def get_ipython(self):
         """Return the currently running IPython instance."""
         return self
