diff --git a/xeofs/models/__init__.py b/xeofs/models/__init__.py
index 49ee8ec..3513b20 100644
--- a/xeofs/models/__init__.py
+++ b/xeofs/models/__init__.py
@@ -2,7 +2,7 @@ from .eof import EOF, ComplexEOF
 from .mca import MCA, ComplexMCA
 from .eeof import ExtendedEOF
 from .opa import OPA
-from .gwpca import GWPCA
+# from .gwpca import GWPCA
 from .rotator_factory import RotatorFactory
 from .eof_rotator import EOFRotator, ComplexEOFRotator
 from .mca_rotator import MCARotator, ComplexMCARotator
