diff --git a/pyproject.toml b/pyproject.toml
index 5701cf1..e9f063d 100644
--- a/pyproject.toml
+++ b/pyproject.toml
@@ -56,7 +56,7 @@ optional-dependencies.all = [
   "geopandas",
   "hda>=2.22",
   "polytope-client>=0.7.4",
-  "pyfdb>=0.1",
+  # "pyfdb>=0.1",
   "pyodc",
 ]
 optional-dependencies.cds = [ "cdsapi>=0.7.2" ]
@@ -75,7 +75,7 @@ optional-dependencies.dev = [
   "nbconvert",
   "nbformat",
   "polytope-client>=0.7.4",
-  "pyfdb>=0.1",
+  # "pyfdb>=0.1",
   "pyodc",
   "pytest",
   "pytest-cov",
@@ -93,7 +93,7 @@ optional-dependencies.docs = [
   "sphinx-tabs",
 ]
 optional-dependencies.ecmwf-opendata = [ "ecmwf-opendata>=0.3.3" ]
-optional-dependencies.fdb = [ "pyfdb>=0.1" ]
+# optional-dependencies.fdb = [ "pyfdb>=0.1" ]
 optional-dependencies.geo = [ "earthkit-geo>=0.2" ]
 optional-dependencies.geopandas = [ "geopandas" ]
 optional-dependencies.geotiff = [ "pyproj", "rasterio", "rioxarray" ]
