diff --git a/pyicon/pyicon_plotting.py b/pyicon/pyicon_plotting.py
index 0373faf..d808401 100644
--- a/pyicon/pyicon_plotting.py
+++ b/pyicon/pyicon_plotting.py
@@ -15,7 +15,7 @@ import cartopy
 import cartopy.crs as ccrs
 from cartopy.mpl.ticker import LongitudeFormatter, LatitudeFormatter
 import cmocean
-from ipdb import set_trace as mybreak
+# from ipdb import set_trace as mybreak
 #from importlib import reload
 from .pyicon_tb import write_dataarray_to_nc
 from .pyicon_tb import identify_grid
diff --git a/pyicon/pyicon_tb.py b/pyicon/pyicon_tb.py
index e9e8040..e024185 100644
--- a/pyicon/pyicon_tb.py
+++ b/pyicon/pyicon_tb.py
@@ -18,7 +18,7 @@ import matplotlib.pyplot as plt
 import matplotlib
 # --- debugging
 #print('mybreak')
-from ipdb import set_trace as mybreak  
+# from ipdb import set_trace as mybreak
 #print('pnadas')
 import pandas as pd
 #print('xarray')
diff --git a/pyicon/quickplots/old_qp_compare_atm_dwd.py b/pyicon/quickplots/old_qp_compare_atm_dwd.py
index 3968a24..bc746f7 100644
--- a/pyicon/quickplots/old_qp_compare_atm_dwd.py
+++ b/pyicon/quickplots/old_qp_compare_atm_dwd.py
@@ -1,6 +1,6 @@
 import sys, glob, os
 import argparse
-from ipdb import set_trace as mybreak
+# from ipdb import set_trace as mybreak
 
 #class DS_attributes(object):
 #  def __init__(self, prefix, output_freq='yearly'):
diff --git a/pyicon/quickplots/old_qp_driver.py b/pyicon/quickplots/old_qp_driver.py
index 2325943..48c2bd0 100644
--- a/pyicon/quickplots/old_qp_driver.py
+++ b/pyicon/quickplots/old_qp_driver.py
@@ -1,6 +1,6 @@
 import sys, glob, os
 import argparse
-from ipdb import set_trace as mybreak
+# from ipdb import set_trace as mybreak
 
 #class DS_attributes(object):
 #  def __init__(self, prefix, output_freq='yearly'):
diff --git a/pyicon/quickplots/pyicon_quickplots.py b/pyicon/quickplots/pyicon_quickplots.py
index 3c24ad3..4714db5 100644
--- a/pyicon/quickplots/pyicon_quickplots.py
+++ b/pyicon/quickplots/pyicon_quickplots.py
@@ -17,7 +17,7 @@ import cartopy.crs as ccrs
 from cartopy.mpl.ticker import LongitudeFormatter, LatitudeFormatter
 import cmocean
 # --- debugging
-from ipdb import set_trace as mybreak  
+# from ipdb import set_trace as mybreak
 import pyicon as pyic
 import xarray as xr
 import cftime
diff --git a/pyicon/quickplots/qp_compare.py b/pyicon/quickplots/qp_compare.py
index 0f34aaf..7500a62 100644
--- a/pyicon/quickplots/qp_compare.py
+++ b/pyicon/quickplots/qp_compare.py
@@ -1,6 +1,6 @@
 import sys, glob, os
 import argparse
-from ipdb import set_trace as mybreak
+# from ipdb import set_trace as mybreak
 import matplotlib
 #if iopts.batch or iopts.slurm:
 #  matplotlib.use('Agg')
diff --git a/pyicon/quickplots/qp_compare_atm_dwd.py b/pyicon/quickplots/qp_compare_atm_dwd.py
index 8a0b499..9c9e527 100644
--- a/pyicon/quickplots/qp_compare_atm_dwd.py
+++ b/pyicon/quickplots/qp_compare_atm_dwd.py
@@ -1,6 +1,6 @@
 import sys, glob, os
 import argparse
-from ipdb import set_trace as mybreak
+# from ipdb import set_trace as mybreak
 
 #class DS_attributes(object):
 #  def __init__(self, prefix, output_freq='yearly'):
diff --git a/scripts/pyic_fig.py b/scripts/pyic_fig.py
index 4d5cec0..70995d4 100755
--- a/scripts/pyic_fig.py
+++ b/scripts/pyic_fig.py
@@ -166,7 +166,7 @@ def main():
     import os
     import sys
     from pathlib import Path
-    from ipdb import set_trace as mybreak  
+    # from ipdb import set_trace as mybreak
     #sys.path.append(f'{Path.home()}/pyicon/')
     import pyicon as pyic  
     print('Done loading modules.')
diff --git a/scripts/pyic_intake.py b/scripts/pyic_intake.py
index a3d2306..87c8529 100755
--- a/scripts/pyic_intake.py
+++ b/scripts/pyic_intake.py
@@ -3,7 +3,7 @@ import os
 import glob
 import yaml
 import argparse
-from ipdb import set_trace as mybreak
+# from ipdb import set_trace as mybreak
 import sys
 from netCDF4 import Dataset
 import pandas as pd
diff --git a/scripts/pyic_sec.py b/scripts/pyic_sec.py
index 87c57d8..804bf8c 100755
--- a/scripts/pyic_sec.py
+++ b/scripts/pyic_sec.py
@@ -146,7 +146,7 @@ def main():
     import os
     import sys
     from pathlib import Path
-    from ipdb import set_trace as mybreak  
+    # from ipdb import set_trace as mybreak
     #sys.path.append(f'{Path.home()}/pyicon/')
     import pyicon as pyic  
     print('Done loading modules.')
